### PR TITLE
Modernize cmake

### PR DESCRIPTION
**Pull request by orzechow**
Some CMake cleanup and refactoring, in preparation to add the #22  as subfolder project.

Based on [Modern CMake book](https://cliutils.gitlab.io/modern-cmake/README.html) and other sources.

### CODE REVIEW

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+###################
+
+find_package(GTest)
+
+
+###########
+## Build ##
+###########
+
+if(GTEST_FOUND)
```
Comment (ll-nick) [inline test/CMakeLists.txt:12]: We _could_ add a warning here if gtest is not found. This CMakeLists should only be configured/built if we explicitly want unit tests.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,33 +1,68 @@
...
+
+
+###############
+## C++ setup ##
+###############
+
+# Only do these if this is the main project, and not if it is included through add_subdirectory
+if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
+  # Require C++17
+  set(CMAKE_CXX_STANDARD 17)
```
Comment (ll-nick) [inline CMakeLists.txt:31]: What if the package including this package sets the cxx standard to e.g. 11?

File: README.md [hunk 3]
```diff
@@ -138,12 +138,12 @@ In order to compile with tests define `BUILD_TESTS=true`
...
-cmake --build .
+cmake --build . -j9
 ```
 
-Run all unit tests:
+Run all unit tests with CTest:
 
 ```bash
-find -executable -type f -name '*-gtest-*' -exec {} \;
+cmake --build . --target test
```
Comment (ll-nick) [inline README.md:147]: :ok_hand:

File: Dockerfile [hunk 4]
```diff
@@ -1,4 +1,4 @@
-FROM ubuntu:22.04
+FROM ubuntu:22.04 as base
```
Comment (ll-nick) [inline Dockerfile:1]:
```suggestion
FROM ubuntu:22.04 AS base
```

File: Dockerfile [hunk 5]
```diff
@@ -23,21 +23,28 @@ RUN git clone https://github.com/KIT-MRT/util_caching.git /tmp/util_caching && \
...
+FROM base AS devel
+
+RUN useradd --create-home --uid 1000 blinky
+USER blinky
+
+WORKDIR /home/blinky/
+
+
+
+FROM base as install
```
Comment (ll-nick) [inline Dockerfile:36]:
```suggestion
FROM base AS install
```

**Review by ll-nick — APPROVED** [review]
Awesome, thanks a lot!

Lots of great changes, I don't have a lot to complain about. :)

File: CMakeLists.txt (hunk 2 above)
Comment (orzechow) [reply CMakeLists.txt:31]: Then, compilation will fail. C++17 is the minimum we require.
Such users can install our package before usage.

But, with this formulation, we allow users to include the package as subdirectory when using newer standards.

File: Dockerfile (hunk 4 above)
Comment (orzechow) [reply Dockerfile:1]: :+1:

File: README.md (hunk 3 above)
Comment (orzechow) [reply README.md:147]: 😀

File: README.md [hunk 6]
```diff
@@ -138,7 +138,7 @@ In order to compile with tests define `BUILD_TESTS=true`
 mkdir -p arbitration_graphs/build
 cd arbitration_graphs/build
 cmake -DBUILD_TESTS=true ..
-cmake --build .
+cmake --build . -j9
```
Comment (ll-nick) [inline README.md:141]: Why 9 btw? :D

Comment (orzechow) [reply README.md:141]: Kind of arbitraty, but at least more than one.

A very rough suggestion was once to use `#threads + 1`, nowadays `2 * #threads` seems to also work for many. Anyhow, it's not too crazy to assume a modern dev has >8 threads available (4 cores with hyperthreading).

**Pull request merged**